packages/artifact/src/upload-http-client.ts

```diff
@@ -0,0 +1,337 @@
+import {debug} from '@actions/core'
+import {HttpClientResponse, HttpClient} from '@actions/http-client/index'
+import {IHttpClientResponse} from '@actions/http-client/interfaces'
+import {
+  ArtifactResponse,
+  CreateArtifactParameters,
+  PatchArtifactSize,
+  UploadResults
+} from './contracts'
+import * as fs from 'fs'
+import {UploadSpecification} from './upload-specification'
+import {UploadOptions} from './upload-options'
+import {URL} from 'url'
+import {
+  createHttpClient,
+  getArtifactUrl,
+  getContentRange,
+  getRequestOptions,
+  isRetryableStatusCode,
+  isSuccessStatusCode
+} from './utils'
+import {
+  getUploadChunkConcurrency,
+  getUploadChunkSize,
+  getUploadFileConcurrency
+} from './config-variables'
+
+/**
+ * Creates a file container for the new artifact in the remote blob storage/file service
+ * @param {string} artifactName Name of the artifact being created
+ * @returns The response from the Artifact Service if the file container was successfully created
+ */
+export async function createArtifactInFileContainer(
+  artifactName: string
+): Promise<ArtifactResponse> {
+  const parameters: CreateArtifactParameters = {
+    Type: 'actions_storage',
+    Name: artifactName
+  }
+  const data: string = JSON.stringify(parameters, null, 2)
+  const artifactUrl = getArtifactUrl()
+  const client = createHttpClient()
+  const requestOptions = getRequestOptions('application/json')
+
+  const rawResponse = await client.post(artifactUrl, data, requestOptions)
+  const body: string = await rawResponse.readBody()
+
+  if (
+    isSuccessStatusCode(rawResponse.message.statusCode) &&
+    body
+  ) {
+    return JSON.parse(body)
+  } else {
+    // eslint-disable-next-line no-console
+    console.log(rawResponse)
+    throw new Error(
+      `Unable to create a container for the artifact ${artifactName}`
+    )
+  }
+}
+
+/**
+ * Concurrently upload all of the files in chunks
+ * @param {string} uploadUrl Base Url for the artifact that was created
+ * @param {SearchResult[]} filesToUpload A list of information about the files being uploaded
+ * @returns The size of all the files uploaded in bytes
+ */
+export async function uploadArtifactToFileContainer(
+  uploadUrl: string,
+  filesToUpload: UploadSpecification[],
+  options?: UploadOptions
+): Promise<UploadResults> {
+  const client = createHttpClient()
+  const FILE_CONCURRENCY = getUploadFileConcurrency()
+  const CHUNK_CONCURRENCY = getUploadChunkConcurrency()
+  const MAX_CHUNK_SIZE = getUploadChunkSize()
+  debug(
+    `File Concurrency: ${FILE_CONCURRENCY}, Chunk Concurrency: ${CHUNK_CONCURRENCY} and Chunk Size: ${MAX_CHUNK_SIZE}`
+  )
+
+  const parameters: UploadFileParameters[] = []
+  let continueOnError = true
+  if (options) {
```

Comment:
This is not future proof - if we add another option in the future, this check will fail.  Would 
` const continueOnError = !!(options && options.continueOnError)` 
work?

Reply:
In `upload-options.ts`, `continueOnError` is a required property though, so as long as there are some `uploadOptions` provided this should work even if some extra properties are added to `uploadOptions`

If there ever are more options added, they should be required and the general flow will be something like this
```
// define default options
if (options){
// update default options to whatever is specified in options
}
```

Reply:
In general, `options` should **not** be required.  Imagine if there are many options available, user should only need to set the ones they care about.

Reply:
Changed `continueOnError` so it is optional in `upload-options.ts` 

Changed the rest to this:
```
  // by default, file uploads will continue if there is an error unless specified differently in the options
  let continueOnError = true
  if (options) {
    if (options.continueOnError === false) {
      continueOnError = false
    }
  }
```
The double bang operators didn't seem very readable so I tried to avoid it. We do also need to support a default value if no `continueOnError` option is specified. Doing all that in a single line is possible but it gets pretty ugly so I think what I have is easy to follow

Reply:
I think this would be more readable as Yang suggested:
```js
const continueOnError = !!(options && options.continueOnError)
```

Reply:
So that one liner would be fine _if_ we didn't care about the default `true` behavior if no options are provided. So it just won't work

Changing to just that one line changes the expected behavior if no options are provided (2 tests start to fail).